__launch_bounds__ on occa-generated hip source called with argument (0).

Hi team,
I have an okl kernel in our code which works well on Nvidia device but same okl when compiled with hip for amd GPUs, gives error (shown in stack below) while JITing. one of the kernels (lets same named kernel1_name)which gets decorated with __launch_bounds__ has (0) arg for launch_bounds and throws error during JIT compilation for obvious reasons. The same kernel on cuda machine gets __launch_bound__(256).
That is how the definition of kernel starts in source.cpp of occa-generated code.
`extern "C" __global__ __launch_bounds__(0) void _occa_kernel1_nameHIP_OKL_5632125855466636082_0( ...)`

```
Compiling [<kernel1_name>HIP_OKL_17413145964116165067]
hipcc --genco    --amdgpu-target=gfx908:sramecc+:xnack- -I<path>/occa/opt/release-intel-hip/include -I<path>/occa/build/release-intel-hip/include tmp/occa/release-intel-hip/cache/c1b591146129fe26/source.cpp -o tmp/occa/release-intel-hip/cache/c1b591146129fe26/1be1d7ef8f0718ab.binary 2>&1
terminate called after throwing an instance of 'occa::exception'
  what():
---[ Error ]--------------------------------------------------------------------
    File     : occa/src/occa/internal/modes/hip/device.cpp
    Line     : 317
    Function : compileKernel
    Message  : Error compiling [<kernel1_name>HIP_OKL_17413145964116165067], Command: [hipcc --genco    --amdgpu-target=gfx908:sramecc+:xnack- -Iocca/opt/release-intel-hip/include -Iocca/build/release-intel-hip/include tmp/occa/release-intel-hip/cache/c1b591146129fe26/source.cpp -o tmp/occa/release-intel-hip/cache/c1b591146129fe26/1be1d7ef8f0718ab.binary 2>&1]
               Output:

               tmp/occa/release-intel-hip/cache/c1b591146129fe26/source.cpp:1244:23: error: 'amdgpu_flat_work_group_size' attribute argument is invalid: min must not be greater than max
               extern "C" __global__ __launch_bounds__(0) void _occa_<kernel1_name>HIP_OKL_17413145964116165067_0(arg1 * P1_in,
                                     ^
               HIP/4.5.0-GCCcore-11.2.0-amd/include/hip/amd_detail/amd_hip_runtime.h:136:50: note: expanded from macro '__launch_bounds__'
                 select_impl_(__VA_ARGS__, launch_bounds_impl1, launch_bounds_impl0, )(__VA_ARGS__)
                                                                ^
               1 error generated when compiling for gfx908.

    Stack
      22 occa::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
      21 occa::hip::device::compileKernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, occa::json const&)
      20 occa::hip::device::buildKernelFromProcessedSource(occa::hash_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, occa::lang::sourceMetadata_t&, occa::lang::sourceMetadata_t&, occa::json const&)
      19 std::_Function_handler<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), occa::launchedModeDevice_t::buildKernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, occa::hash_t, bool, occa::json const&)::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
      18 occa::io::stageFiles(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool, std::function<bool (std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)>)
      17 occa::io::stageFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::function<bool (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>)
      16 occa::launchedModeDevice_t::buildKernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, occa::hash_t, bool, occa::json const&)
      15 occa::launchedModeDevice_t::buildKernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, occa::hash_t, occa::json const&)
      14 occa::device::buildKernel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, occa::json const&) const
      13 occa/opt/release-intel-hip/lib/libocca.so(occaDeviceBuildKernel+0x531)
      12 ./TestSimulation()
      11 ./TestSimulation()
```

--- a/src/occa/internal/lang/builtins/attributes/maxInnerDims.cpp
+++ b/src/occa/internal/lang/builtins/attributes/maxInnerDims.cpp
@@ -41,7 +41,7 @@ namespace occa {
             error = !value.isInteger();
             if(!error) {
               int intValue = value;
-              error = (intValue < 0);
+              error = !(0 < intValue);
             }
           }
           if(error) {

--- a/src/occa/internal/lang/modes/oklForStatement.cpp
+++ b/src/occa/internal/lang/modes/oklForStatement.cpp
@@ -44,6 +44,19 @@ namespace occa {
           && hasValidCheck()
           && hasValidUpdate()
         );
+
+        if(valid) {
+          exprNode* loop_range_node = getIterationCount();
+          if (loop_range_node->canEvaluate()) {
+            int loop_range = loop_range_node->evaluate();
+            if (!(0 < loop_range)) {
+              valid = false;
+              if(printErrors_)
+                forSmnt_.printError("OKL for loop is empty or infinite!");
+            }
+          }
+          delete loop_range_node;
+        }
       }
 
       bool oklForStatement::isValid() {
@@ -263,14 +276,17 @@ namespace occa {
         if (!valid) {
           return NULL;
         }
-
         exprNode *initInParen = initValue->wrapInParentheses();
-        exprNode *count = (
-          new binaryOpNode(iterator->source,
-                           positiveUpdate ? op::sub : op::add,
-                           *checkValue,
-                           *initInParen)
-        );
+        
+        //If incrementing, assume loop bound is large than initial value
+        //If decrementing: assume initial value is larger than loop bound
+        exprNode *smaller = (positiveUpdate) ? initInParen : checkValue;
+        exprNode *larger  = (positiveUpdate) ? checkValue : initInParen;
+        exprNode *count = (new binaryOpNode(iterator->source,
+                           op::sub,
+                           *larger,
+                           *smaller)
+        );  
         delete initInParen;
 
         if (checkIsInclusive) {
@@ -278,29 +294,34 @@ namespace occa {
 
           exprNode *countWithInc = (
             new binaryOpNode(iterator->source,
-                             positiveUpdate ? op::sub : op::add,
-                             *count,
-                             inc)
+                             op::add,
+                             inc,
+                             *count)
           );
           delete count;
           count = countWithInc;
         }
 
         if (updateValue) {
+          // ceil(n/m) = floor((n+m-1)/m)
           exprNode *updateInParen = updateValue->wrapInParentheses();
+          binaryOpNode countPlusUpdate(
+            iterator->source,
+            op::add,
+            *count,
+            *updateInParen
+          );
 
           primitiveNode one(iterator->source, 1);
-          binaryOpNode boundCheck(iterator->source,
-                                  positiveUpdate ? op::add : op::sub,
-                                  *count,
-                                  *updateInParen);
-          binaryOpNode boundCheck2(iterator->source,
-                                   positiveUpdate ? op::sub : op::add,
-                                   boundCheck,
-                                   one);
-          exprNode *boundCheckInParen = boundCheck2.wrapInParentheses();
+          binaryOpNode countPlusUpdateMinusOne(
+            iterator->source,
+            op::sub,
+            countPlusUpdate,
+            one
+          );
+          exprNode *boundCheckInParen = countPlusUpdateMinusOne.wrapInParentheses();
 
-          exprNode *countWithUpdate = (
+          exprNode *ceilingQuotient = (
             new binaryOpNode(iterator->source,
                              op::div,
                              *boundCheckInParen,
@@ -309,7 +330,7 @@ namespace occa {
           delete count;
           delete updateInParen;
           delete boundCheckInParen;
-          count = countWithUpdate;
+          count = ceilingQuotient;
         }
 
         return count;

--- a/tests/src/internal/lang/modes/okl.cpp
+++ b/tests/src/internal/lang/modes/okl.cpp
@@ -42,12 +42,18 @@ void testOKLLoopExists();
 void testProperOKLLoops();
 void testInnerInsideOuter();
 void testSameInnerLoopCount();
+void testEmptyLoops();
+void testInfiniteLoops();
+void testMaxInnerDims();
 
 void testLoops() {
   testOKLLoopExists();
   testProperOKLLoops();
   testInnerInsideOuter();
   testSameInnerLoopCount();
+  testEmptyLoops();
+  testInfiniteLoops();
+  testMaxInnerDims();
 }
 
 void testOKLLoopExists() {
@@ -159,6 +165,66 @@ void testSameInnerLoopCount() {
     "    for (int i = 0; i < 2; ++i; @inner) {\n"
     "      for (int i2 = 0; i2 < 2; ++i2; @inner) {\n"
     "      }\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+  );
+}
+
+void testEmptyLoops() {
+
+  parseBadOKLSource(
+    "@kernel void foo() {\n"
+    "  for (int o = 0; o < 2; ++o; @outer) {\n"
+    "    for (int i = 10; i < 2; i+=3; @inner) {\n"
+    "        int k = i + j;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+  );
+
+  parseBadOKLSource(
+    "@kernel void foo() {\n"
+    "  for (int o = 0; o < 2; ++o; @outer) {\n"
+    "    for (int i = 10; i > 11; i-=3; @inner) {\n"
+    "        int k = i + j;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+  );
+}
+
+void testInfiniteLoops() {
+
+  parseBadOKLSource(
+    "@kernel void foo() {\n"
+    "  for (int o = 0; o < 2; ++o; @outer) {\n"
+    "    for (int i = 0; i < 2; i-=5; @inner) {\n"
+    "        int k = i + j;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+  );
+
+  parseBadOKLSource(
+    "@kernel void foo() {\n"
+    "  for (int o = 0; o < 2; ++o; @outer) {\n"
+    "    for (int i = 2; i > 2=0; i+=5; @inner) {\n"
+    "        int k = i + j;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+  );
+
+}
+
+void testMaxInnerDims() {
+  parseBadOKLSource(
+    "@kernel void foo(const int& N) {\n"
+    "  @max_inner_dims(0)\n"
+    "  for (int o = 0; o < 2; ++o; @outer) {\n"
+    "    for (int i = 0; i < N; ++i; @inner) {\n"
+    "        int k = i + j;\n"
     "    }\n"
     "  }\n"
     "}\n"

